packages/ui/primitives/document-flow/add-signature.tsx

```diff
@@ -0,0 +1,284 @@
+import {
+  SinglePlayerModeCustomTextField,
+  SinglePlayerModeSignatureField,
+} from './single-player-mode-fields';
+
+export type AddSignatureFormProps = {
+  defaultValues?: TAddSignatureFormSchema;
+  documentFlow: DocumentFlowStep;
+  fields: FieldWithSignature[];
+  numberOfSteps: number;
+  onSubmit: (_data: TAddSignatureFormSchema) => Promise<void> | void;
+  requireName?: boolean;
+  requireSignature?: boolean;
+};
+
+export const AddSignatureFormPartial = ({
+  defaultValues,
+  documentFlow,
+  fields,
+  numberOfSteps,
+  onSubmit,
+  requireName = false,
+  requireSignature = true,
+}: AddSignatureFormProps) => {
+  // Refined schema which takes into account whether to allow an empty name or signature.
+  const refinedSchema = ZAddSignatureFormSchema.superRefine((val, ctx) => {
+    if (requireName && val.name.length === 0) {
+      ctx.addIssue({
+        path: ['name'],
+        code: 'custom',
+        message: 'Name is required',
+      });
+    }
+
+    if (requireSignature && val.signature.length === 0) {
+      ctx.addIssue({
+        path: ['signature'],
+        code: 'custom',
+        message: 'Signature is required',
+      });
+    }
+  });
+
+  const form = useForm<TAddSignatureFormSchema>({
+    resolver: zodResolver(refinedSchema),
+    defaultValues: defaultValues ?? {
+      name: '',
+      email: '',
+      signature: '',
+    },
+  });
+
+  /**
+   * A local copy of the provided fields to modify.
+   */
+  const [localFields, setLocalFields] = useState(
+    fields.map((field) => {
+      let inserted = field.inserted;
+      let customText = field.customText;
+
+      switch (field.type) {
+        case FieldType.NAME:
+          inserted = form.getValues('name').length > 0;
+          break;
+        case FieldType.EMAIL:
+          inserted = form.getValues('email').length > 0;
+          break;
+        case FieldType.SIGNATURE:
+          inserted = form.getValues('signature').length > 0;
+          break;
+        case FieldType.DATE:
+          inserted = true;
+
+          // Todo: Extract.
+          customText = DateTime.now().toFormat('yyyy-MM-dd hh:mm a');
+          break;
+        default:
+          inserted = true;
+          break;
+      }
+
+      return { ...field, inserted, customText };
+    }),
+  );
+
+  return (
+    <Form {...form}>
+      <fieldset className="flex h-full flex-col" disabled={form.formState.isSubmitting}>
+        <DocumentFlowFormContainerContent>
+          <div className="space-y-4">
+            <FormField
+              control={form.control}
+              name="email"
+              render={({ field }) => (
+                <FormItem>
+                  <FormLabel required>Email</FormLabel>
+                  <FormControl>
+                    <Input
+                      className="bg-background"
+                      type="email"
+                      autoComplete="email"
+                      {...field}
+                      onBlur={() => {
```

Comment:
😬

Reply:
Partially extracted the items out, still using inline functions to trigger the native event on the field.